tests/unit/test_endpoints_v2.py

```diff
@@ -0,0 +1,555 @@
+    path = os.path.join(
+        os.path.dirname(__file__), 'data', 'endpoints', 'rules', 'ruleset.json'
+    )
+    with open(path) as f:
+        data = f.read()
+        return json.loads(data)
+
+
+@pytest.fixture
+def endpoint_provider(ruleset_dict, partitions):
+    return EndpointProvider(ruleset_dict, partitions)
+
+
+@pytest.fixture
+def ruleset(endpoint_provider):
+    return endpoint_provider.ruleset
+
+
+@pytest.fixture
+def endpoint_rule(partitions):
+    return EndpointRule(
+        endpoint=endpoint_dict(),
+        conditions=[
+            parse_arn_func(),
+            {
+                'fn': 'not',
+                'argv': [string_equals_func()],
+            },
+            {
+                'fn': 'aws.partition',
+                'argv': [region_ref()],
+                'assign': 'PartitionResults',
+            },
+        ],
+        partitions=partitions,
+    )
+
+
+def generated_ruleset_testcases(case_type):
+    filenames = [
+        'aws-region',
+        'default-values',
+        # 'deprecated-param',
+        'eventbridge',
+        'fns',
+        # 'get-attr-type-interface',
+        'headers',
+        'local-region-override',
+        # 'minimal-ruleset',
+        'parse-arn',
+        'parse-url',
+        'substring',
+        'uri-encode',
+        'valid-hostlabel',
+        # 'valid-ruleset'
+    ]
+
+    base_path = os.path.join(os.path.dirname(__file__), 'data', 'endpoints')
+    for name in filenames:
+
+        with open(os.path.join(base_path, 'rules', f'{name}.json')) as f:
+            data = f.read()
+            ruleset = json.loads(data)
+        with open(os.path.join(base_path, 'test_cases', f'{name}.json')) as f:
+            data = f.read()
+            tests = json.loads(data)
+
+        for test in tests['testCases']:
+            input_params = test['params']
+            expected_object = test['expect']
+            if case_type in expected_object:
+                yield (ruleset, input_params, expected_object)
+
+
+@pytest.mark.parametrize(
+    "ruleset,input_params,expected_object",
+    generated_ruleset_testcases('error'),
+)
+def test_endpoint_resolution_raises(
+    partitions, ruleset, input_params, expected_object
+):
+    endpoint_provider = EndpointProvider(ruleset, partitions)
+    with pytest.raises(EndpointResolutionError) as err:
+        endpoint_provider.resolve_endpoint(**input_params)
+        assert err.msg == expected_object['error']
+
+
+@pytest.mark.parametrize(
+    "ruleset,input_params,expected_object", generated_ruleset_testcases('url')
+)
+def test_endpoint_resolution(
+    partitions, ruleset, input_params, expected_object
+):
+    endpoint_provider = EndpointProvider(ruleset, partitions)
+    endpoint = endpoint_provider.resolve_endpoint(**input_params)
+    assert endpoint.url == expected_object['url']
+    assert endpoint.properties == expected_object.get('properties', {})
+    assert endpoint.headers == expected_object.get('headers', {})
```

Comment:
Do we have test cases where properties/headers aren't present?

Reply:
there are generated test cases without these keys present yes.